web/app/js/components/Tap.jsx

```diff
@@ -131,6 +129,60 @@ class Tap extends React.Component {
     }, {});
   }
 
+  parseTapResult = data => {
+    let d = JSON.parse(data);
+
+    switch (d.proxyDirection) {
+      case "INBOUND":
+        d.tls = _.get(d, "sourceMeta.labels.tls", "");
+        break;
+      case "OUTBOUND":
+        d.tls = _.get(d, "destinationMeta.labels.tls", "");
+        break;
+      default:
+        // too old for TLS
+    }
+
+    if (_.isNil(d.http)) {
+      this.setState({ error: "Undefined request type"});
+    } else {
+      if (!_.isNil(d.http.requestInit)) {
+        d.eventType = "req";
+        d.id = `${_.get(d, "http.requestInit.id.base")}:${_.get(d, "http.requestInit.id.stream")} `;
+      } else if (!_.isNil(d.http.responseInit)) {
+        d.eventType = "rsp";
+        d.id = `${_.get(d, "http.responseInit.id.base")}:${_.get(d, "http.responseInit.id.stream")} `;
+      } else if (!_.isNil(d.http.responseEnd)) {
+        d.eventType = "end";
+        d.id = `${_.get(d, "http.responseEnd.id.base")}:${_.get(d, "http.responseEnd.id.stream")} `;
+      }
+    }
+
+    return d;
+  }
+
+  indexTapResult = data => {
+    // keep an index of tap request rows by id. this allows us to collate
+    // requestInit/responseInit/responseEnd into one single table row,
+    // as opposed to three separate rows as in the CLI
+    let resultIndex = this.state.tapResultsById;
+    let d = this.parseTapResult(data);
+
+    if (_.isNil(resultIndex[d.id])) {
+      // don't let tapResultsById grow unbounded
+      if (_.size(resultIndex) > this.state.maxLinesToDisplay) {
+        resultIndex = {};
```

Comment:
This is totally fine for now, but the net effect of this reassignment is that once the table grows to 40 rows, we truncate to 1 row in the UI, which is a bit jarring. It might be better if we just dropped the oldest row. But that would require switching to a different data structure to track the oldest row, so I'd be fine if you want to do that separately or not at all.

Reply:
yeah, agreed that the current transition is not ideal. It's not a huge deal to drop the oldest row, I'd be in favour of that!